ext/http/network_buffered_stream.rs

```diff
@@ -0,0 +1,284 @@
+use std::task::Poll;
+use tokio::io::AsyncRead;
+use tokio::io::AsyncWrite;
+use tokio::io::ReadBuf;
+
+const MAX_PREFIX_SIZE: usize = 256;
+
+pub struct NetworkStreamPrefixCheck<S: AsyncRead + Unpin> {
+  io: S,
+  prefix: &'static [u8],
+  buffer: [MaybeUninit<u8>; MAX_PREFIX_SIZE * 2],
+}
+
+impl<S: AsyncRead + Unpin> NetworkStreamPrefixCheck<S> {
+  pub fn new(io: S, prefix: &'static [u8]) -> Self {
+    debug_assert!(prefix.len() < MAX_PREFIX_SIZE);
+    Self {
+      io,
+      prefix,
+      buffer: [MaybeUninit::<u8>::uninit(); MAX_PREFIX_SIZE * 2],
+    }
+  }
+
+  // Returns a [`NetworkBufferedStream`], rewound with the bytes we read to determine what
+  // type of stream this is.
+  pub async fn match_prefix(
+    self,
+  ) -> io::Result<(bool, NetworkBufferedStream<S>)> {
+    let mut buffer = self.buffer;
+    let mut readbuf = ReadBuf::uninit(&mut buffer);
+    let mut io = self.io;
+    let prefix = self.prefix;
+    loop {
+      enum State {
+        Unknown,
+        Matched,
+        NotMatched,
+      }
+
+      let state = poll_fn(|cx| {
+        let filled_len = readbuf.filled().len();
+        let res = ready!(Pin::new(&mut io).poll_read(cx, &mut readbuf));
+        if let Err(e) = res {
+          return Poll::Ready(Err(e));
+        }
+        let filled = readbuf.filled();
+        let new_len = filled.len();
+        if new_len == filled_len {
+          // Empty read, no match
+          return Poll::Ready(Ok(State::NotMatched));
+        } else if new_len < prefix.len() {
+          // Read less than prefix, make sure we're still matching the prefix (early exit)
+          if !prefix.starts_with(filled) {
+            return Poll::Ready(Ok(State::NotMatched));
+          }
+        } else if new_len >= prefix.len() {
+          // We have enough to determine
+          if filled.starts_with(prefix) {
+            return Poll::Ready(Ok(State::Matched));
+          } else {
+            return Poll::Ready(Ok(State::NotMatched));
+          }
+        }
+
+        Poll::Ready(Ok(State::Unknown))
+      })
+      .await?;
+
+      match state {
+        State::Unknown => continue,
+        State::Matched => {
+          let initialized_len = readbuf.filled().len();
+          return Ok((
+            true,
+            NetworkBufferedStream::new(io, buffer, initialized_len),
+          ));
+        }
+        State::NotMatched => {
+          let initialized_len = readbuf.filled().len();
+          return Ok((
+            false,
+            NetworkBufferedStream::new(io, buffer, initialized_len),
+          ));
+        }
+      }
+    }
+  }
+}
+
+pub struct NetworkBufferedStream<S: AsyncRead + Unpin> {
+  io: S,
+  initialized_len: usize,
+  prefix_offset: usize,
+  prefix: [MaybeUninit<u8>; MAX_PREFIX_SIZE * 2],
+  prefix_read: bool,
+}
+
+impl<S: AsyncRead + Unpin> NetworkBufferedStream<S> {
+  fn new(
+    io: S,
+    prefix: [MaybeUninit<u8>; MAX_PREFIX_SIZE * 2],
+    initialized_len: usize,
+  ) -> Self {
```

Comment:
I feel like strictly speaking this should be `unsafe`, because you are delegating responsibility of guaranteeing that prefix is initialized to `initialized_len` to the caller.

Not the end of the world though, considering that this method is not `pub`.

Reply:
I considered that as well, but this is "interior" unsafe like you said. This is a bit of a challenging API as we can't pass an initialized slice (because that needs the backing data) and we don't want to copy the array eagerly. The two classes are quite closely intertwined. I can add a note to that and suggest a future visitor to this section of the code consider merging them.